### PR TITLE
ci: split elyra matrix entry per test, pre-pull images to fix TrustyAI spawn timeout

### PR DESCRIPTION
**Pull request by jiridanek**
## Summary

Splits `tests/Tests/0500__ide/0502__ide_elyra.robot`'s two Robot Framework test cases into
separate CI matrix jobs (`sanity`, `tier1`) so each can carry its own optional
`prepull_images`/`extra_robot_args`, then designs and iteratively hardens an in-cluster
DaemonSet-based image pre-pull mechanism to fix the TrustyAI workbench's cold-pull timeout
(#67) — eventually extending pre-pull to every other matrix job's own workbench/runtime images
once the mechanism proved reliable.

## Root cause

TrustyAI (and other large, multi-GB workbench/runtime) images can take several minutes to
cold-pull under CI runner I/O load — longer than the robot test's fixed workbench-spawn
timeout, failing the test for reasons unrelated to the feature under test. Fixed by pre-pulling
images into the kind node *before* the robot test needs them, via an in-cluster DaemonSet
mechanism (not `kind load docker-image`, which turned out to be even slower than a plain
registry pull and eventually failed outright under disk pressure — see Trajectory). The
pre-pull mechanism itself went through several designs before landing on one that's both fast
and doesn't risk starving `deploy.py`'s own image pulls — two separate starvation failure
modes were found and fixed along the way (see Trajectory).

## Related issue

Mitigates #67.

## Test plan

- [x] Verified the `include:`-only matrix conversion produces the expected job count/display
      names (no accidental job merging/collision from GitHub Actions' include+axis semantics)
- [x] Verified the elyra sanity/tier1 split produces unique `upload-artifact` names (no
      collision from both entries sharing the same underlying `.robot` file)
- [x] All 17 matrix entries' `prepull_images` lists were derived from actual `podss.yaml`/
      `eventss.yaml` in cluster-logs artifacts of full CI runs, not guessed from test source —
      cross-checked across multiple independent runs for stability
- [x] Multiple full 18-job CI runs plus targeted `gh run rerun --job` benchmark loops (full
      timing tables in `pulling_timings.md`, kept locally/untracked as a working investigation
      doc) confirm: (a) zero new regressions vs. the pre-change baseline — the same 4
      pre-existing, unrelated jobs fail identically before and after every change in this PR;
      (b) the final two-phase pre-pull design keeps `deploy.py` near its undisturbed baseline
      duration (~250-280s) across 5 consecutive samples, with no repeat of either starvation
      failure mode that earlier designs hit
- [ ] CI (this PR's own checks)

<details>
<summary>Plan</summary>

Original approved plan (since substantially extended through implementation - see Trajectory
for what changed and why):

Split the single "elyra" matrix entry (which today runs the whole `.robot` file containing 2
independent Robot Test Cases) into one matrix entry per Test Case, so each can carry two new
optional per-entry fields: a list of images to pre-pull, and extra robot CLI args to filter
which test in the file actually runs. Add `include:`-only matrix entries (not a flat
`test_case:` axis + `include:` overlay) to sidestep GitHub Actions' ambiguous include/axis merge
semantics when two entries need to share the same `test_case` value. Fix the artifact-name
collision this split introduces (both entries derive the same name from the shared filename) via
a new `artifact_suffix` field. Pre-pull the Tier1 entry's needed images (PyTorch, TensorFlow,
TrustyAI, datascience runtime) via `kind load docker-image` before the robot test runs, reusing
the retry-loop idiom already used for the istio pilot image pre-pull. Thread `extra_robot_args`
through to `run_robot_test.sh`'s existing `--extra-robot-args` passthrough. Give the split
entries a distinct job display name via a `name:` override using `artifact_suffix`.

Research conclusion at plan time on whether to use a more "robust" DevSpaces-like DaemonSet
image-puller instead of `kind load docker-image`: **no, `kind load` is sufficient** — the
DaemonSet approach (`che-incubator/kubernetes-image-puller`, RHOAIENG-56335) only helps on
multi-node autoscaling, where a racing image-puller pod and workload pod land on a new node
together; this kind cluster is one static node that already exists before the test runs, so
`kind load` should be strictly sufficient. **This conclusion was later proven wrong in
practice** (see Trajectory) — `kind load` itself turned out to be the slower, less reliable
mechanism, and the PR ended up building the DaemonSet approach after all.

</details>

<details>
<summary>Trajectory</summary>

**Phase 1 — matrix split + `kind load`-based pre-pull (original plan, commits `dd31a40`,
`fbedda2`):**

1. Investigated a failing elyra CI run and found the templated Tier1 test's TrustyAI workbench
   image (~6GB) sometimes takes ~5 minutes to cold-pull under CI I/O load, exceeding the robot
   test's fixed 5-minute workbench-spawn timeout (baseline evidence from a live CI run: pull
   times datascience 34.5s/1.8GB, pytorch 2m20s/10GB, tensorflow 1m43s/9.3GB, trustyai
   anomalously slow at 5m0.4s/6GB).
2. Confirmed the file's second Robot Test Case is templated over 3 workbench images and runs
   independently of the first (Sanity) test case, both currently forced into one matrix entry.
3. Designed and implemented the `include:`-only matrix split (one entry per Test Case, new
   `artifact_suffix`/`extra_robot_args`/`prepull_images` fields), the artifact-name-collision
   fix, the `extra_robot_args` passthrough, and a `kind load docker-image` pre-pull step for the
   Tier1 entry's 4 images, guarded by `if: matrix.prepull_images` truthiness so it's a no-op for
   every other entry.
4. Pulled `workbench_branch` out of the per-entry matrix into a job-level `env:
   WORKBENCH_BRANCH` (user request: it never actually varies between entries, no reason to
   repeat it ~18 times) — required removing it from the job's `name:` expression too, since
   GitHub Actions doesn't expose the `env` context to `jobs.<job_id>.name` (a restriction
   discovered the hard way via a CI validation failure, fixed in `a593faa`).

**Phase 2 — `kind load` turns out to be the wrong call; replaced with an in-cluster DaemonSet
(commits `ad5aa43`, `ad5b717`, `c15c4c7`, `40f5a45`, `714a9ac`, `0aa873d`):**

5. The first CI run under the new pre-pull step still failed: `kind load docker-image` (which
   internally does `docker save` piped into the kind node container) took far longer than the
   registry pull itself — 7m14s just for the PyTorch image's `kind load`, on top of its own
   ~2m45s `docker pull` — and eventually failed outright with a `docker save` I/O error,
   consistent with the runner's disk filling up from storing the same image twice (once on the
   GHA runner's host Docker daemon, once duplicated into the kind node's containerd). This
   directly falsified the plan's research conclusion that `kind load` would be sufficient.
6. Replaced the mechanism with an in-cluster DaemonSet approach after all — the same
   `che-incubator/kubernetes-image-puller` pattern the original plan had considered and set
   aside: each container runs `sleep 720h` on the target image, so containerd pulls it directly
   (no host-side double-pull) and a real running container keeps kubelet's image GC from
   evicting it. Reused this same pattern for the pre-existing istio pilot image pre-pull too.
7. First DaemonSet placement was *before* `deploy.py` (matching istio's pre-pull position) —
   this caused a **first starvation failure**: the DaemonSet's ~35GB of large image pulls
   monopolized kubelet's default `serializeImagePulls: true` FIFO queue, starving
   cert-manager's own tiny, normally-instant pulls badly enough that `deploy.py`'s 5-minute
   `kubectl wait ... cert-manager` timed out. Fixed by moving DaemonSet creation to *after*
   `deploy.py` instead (the blocking wait for it still deferred to just before the robot test
   runs, so pulling still overlaps with Poetry/oc-login/test-variables setup).
8. User asked whether the DaemonSet could be deprioritized instead of reordered; explored
   `PriorityClass` and initContainer-based serialization, but initContainers exit after running
   and lose kubelet's GC protection (user caught this), so simple reordering was kept as the
   fix.
9. Added visibility: streamed `kubectl get events -n kube-system --watch` in the background
   during the wait step, printing elapsed time, after the user asked for progress output beyond
   `kubectl rollout status`'s silent pod-count reporting.
10. Converted the bash-based pre-pull/wait steps to `shell: python {0}` (existing repo
    convention for building k8s manifests as Python dicts + `json.dumps()`) per user request.

**Phase 3 — discovering kind's kubelet pull concurrency, and single-pod vs. per-image
DaemonSets (commits `1a5b52e`, `0aa873d`, plus `components/00-kind-cluster.yaml`):**

11. Investigated kind's/kubelet's image-pull parallelism defaults; confirmed via
    `kubernetes/kubernetes` release-1.31 source that `serializeImagePulls` defaults to `true`
    and `maxParallelImagePulls` defaults to `nil` (mutually exclusive). Added a
    `kubeadmConfigPatches` `KubeletConfiguration` override in `components/00-kind-cluster.yaml`
    setting `serializeImagePulls: false` / `maxParallelImagePulls: 3` to allow concurrent pulls.
12. First CI run under this setting showed only **one** `Pulling` event at a time despite the
    new setting — root-caused to kubelet's pod worker starting a single pod's own containers
    strictly sequentially, *regardless* of `maxParallelImagePulls` (that setting only governs
    concurrency *across* pods). The single-DaemonSet-with-N-sibling-containers shape used since
    Phase 2 made the new concurrency setting a no-op.
13. Split into **one DaemonSet per image** instead (N separate single-container DaemonSets) so
    `maxParallelImagePulls` could actually apply across pods. Confirmed via the event stream:
    3 of 4 images started pulling within the same second, with the 4th visibly queuing and
    starting only once one of the first 3 finished — first hard confirmation the setting works,
    just not with one pod per N containers.

**Phase 4 — discovering the tier1 image list was incomplete; expanding pre-pull to every
matrix job (commits `2d0a963`, `824e8f6`):**

14. On request, read through actual CI logs to find every image the elyra tier1 test uses.
    Console logs alone were insufficient (the pre-pull wait step's event watcher is scoped to
    `kube-system`, never seeing the per-test `elyra-test-NNNN` namespace); had to download a
    passing run's `cluster-logs` artifact and read `podss.yaml`/`eventss.yaml` directly. Found
    that the per-project DSPA this test creates (via its own Suite Setup, not `deploy.py`) also
    cold-pulls its whole control-plane stack (`ds-pipelines-api-server`, `-persistenceagent`,
    `-scheduledworkflow`, `-argo-workflowcontroller`, `-argo-argoexec`, `-driver`, `-launcher`,
    `-runtime-generic`, `mlmd-grpc-server`), a `mariadb` DB backend (since this repo's earlier
    Kyverno fix forces `pipelineStore: database`), and the `workaround` Suite-Setup workbench's
    minimal image — none of which were being pre-pulled. Added all of them to tier1's list
    (later trimmed back down by the user to just the multi-GB images that actually carry
    timeout risk, once the two-phase design made even the smaller ones' contribution
    marginal).
15. Extended the same discovery method (bulk-downloading a full 18-job run's cluster-logs
    artifacts, grepping `eventss.yaml` for pull events) to every *other* matrix job, adding each
    its own `prepull_images` list. Two special cases: `multiple-image-tags.robot` specifically
    exercises the previous release's (n-1, py311) images, pinned by `@sha256` digest since
    that's how the notebooks repo itself pins n-1 (no moving tag to track); `custom-image.robot`
    imports a hardcoded BYON image unrelated to the `WORKBENCH_BRANCH` release train.
    Verified zero regressions by comparing a full run against the pre-change baseline and
    confirming the exact same 4 pre-existing, unrelated jobs fail in both.

**Phase 5 — `maxParallelImagePulls` bump, a second starvation failure, and the two-phase
redesign (commit `3498be5`):**

16. User asked to try `maxParallelImagePulls: 6` and move the per-image DaemonSet creation back
    to *before* `deploy.py` (to overlap pulling with more of the job, on the theory that with
    real parallelism now available, `deploy.py`'s small pulls could get a fair concurrent slot
    instead of queuing strictly behind the large ones).
17. Ran a 5-sample benchmark of this configuration (`gh run rerun --job` targeting just the
    tier1 job, 4 successes + 1 failure): successes averaged 1312s total, faster than every prior
    configuration — but the failure was a **second, different starvation failure**:
    `deploy.py`'s `kubectl wait ... argocd --timeout 180s` timed out because ArgoCD's own image
    pull didn't get admitted into kubelet's pull queue until ~4.5 minutes in, stuck behind a
    burst of 15 background pulls that all arrived first (a 1-in-5 measured failure rate).
18. Discussed several fixes: lowering `maxParallelImagePulls` back down (rejected — the
    background pods still arrive first regardless of slot count, so fewer slots only makes
    `deploy.py`'s wait *longer*, not safer); raising it further (untested, trades queue
    starvation for raw bandwidth contention); splitting `deploy.py` into two steps so its own
    pulls get a first-mover queue advantage (rejected as too invasive for the payoff).
19. Landed on a **two-phase design** instead: phase 1 is a single DaemonSet with one pod and N
    sibling containers (deliberately reusing the "wrong" shape from Phase 3, since kubelet's
    strictly-sequential per-pod container starts now work *for* this purpose) fired
    fire-and-forget before `deploy.py` with no wait step — it only ever occupies **one**
    `maxParallelImagePulls` admission slot no matter how many images are listed, so `deploy.py`'s
    own pulls always find a free slot. Phase 2 is the existing one-DaemonSet-per-image approach,
    now fired *after* `deploy.py` (safe once its readiness waits are done), with the existing
    blocking wait unchanged in position. Re-requesting an image phase 1 already pulled is a fast
    no-op (content-addressed image store).
20. Also parametrized the ~24 repeated `2025b-v1.36` tag literals across the matrix behind a new
    `env.NOTEBOOK_IMAGE_TAG`, substituted at runtime via a `TAG` placeholder in each
    `prepull_images` entry (GitHub Actions matrix values can't reference the `env` context
    directly).
21. Benchmarked the two-phase design across two separate pushes (5 total tier1-job samples: 1
    initial + 4 `gh run rerun --job` reruns on the first push, then after a rebase onto `main`
    and a force-push, 1 more initial + 2 more reruns): `deploy.py` consistently landed back near
    its undisturbed baseline duration (~250-282s, vs. 347-515s under the risky all-before
    design), with zero repeats of either starvation failure mode across all runs. The only
    failure seen was an unrelated, pre-existing Robot Framework flake ("Saving completed" text
    not appearing within a 5-second window), confirmed independent of pre-pull state (the
    pre-pull steps had already succeeded well before that failure occurred).

**Split off into separate, independent PRs along the way** (not part of this PR's diff):
- #73 (merged) — gave each of the 3 CI workflows a short top-level `name:` for easier scanning,
  requested and implemented as a standalone change off `main`.
- #74 (open) — fixed all 3 workflows' istio pre-pull steps hardcoding
  `docker.io/istio/pilot:1.25.1` while `components/deploy.py`'s actual `ISTIO_VERSION` had moved
  to `1.26.2`, found incidentally while reviewing this PR's workflow file.

Full per-run timing tables, the specific failure logs/error messages for every diagnosed
failure, and the exact benchmark methodology are in `pulling_timings.md` (kept as a local,
untracked working document rather than committed repo documentation).

</details>

## Summary by CodeRabbit

* **Tests**
  * Improved automated test execution with faster environment preparation and more reliable readiness monitoring.
  * Expanded test coverage for Elyra scenarios, including dedicated sanity and Tier 1 validation.
  * Test results and artifacts are now organized more clearly for easier review and troubleshooting.

* **Chores**
  * Improved local test-cluster performance by enabling parallel image downloads.

### CODE REVIEW

File: .github/workflows/rhoai-in-kind-with-ods-ci.yaml [hunk 1]
```diff
@@ -11,34 +11,45 @@ permissions: {}
 
 jobs:
   rhoai-in-kind:
+    name: >-
+      rhoai-in-kind (${{ env.WORKBENCH_BRANCH }}, ${{ matrix.test_case }}${{ matrix.artifact_suffix && format(' [{0}]', matrix.artifact_suffix) || '' }})
```
[the remark this answers is not in the document]
Comment (jiridanek) [reply .github/workflows/rhoai-in-kind-with-ods-ci.yaml]: @coderabbitai recheck, fixed
